packages/adblocker/src/engine/engine.ts

```diff
@@ -697,23 +753,44 @@ export default class FilterEngine extends EventEmitter<
     url,
     hostname,
     domain,
+
+    reference,
   }: {
     url: string;
     hostname: string;
     domain: string | null | undefined;
+
+    reference?: any | undefined;
   }): HTMLSelector[] {
     const htmlSelectors: HTMLSelector[] = [];
 
     if (this.config.enableHtmlFiltering === false || this.config.loadCosmeticFilters === false) {
       return htmlSelectors;
     }
 
-    const rules = this.cosmetics.getHtmlRules({
+    const { rules, matches, exceptions } = this.cosmetics.getHtmlRules({
       domain: domain || '',
       hostname,
       isFilterExcluded: this.isFilterExcluded.bind(this),
     });
 
+    const context: CosmeticFilterMatchingContext = {
+      url,
+      hostname,
+      domain,
+
+      matchType: FilterType.COSMETIC,
+      reference,
+    };
+
+    for (const match of matches) {
+      this.emit('filter-matched', match, context);
+
+      if (exceptions.has(match)) {
+        this.emit('filter-matched', exceptions.get(match)!, context);
+      }
+    }
```

Comment:
Can we do that only if there is any event emitter registered? Otherwise that seems like unnecessary overhead.